pkg/apis/batch/v1/types.go

```diff
@@ -74,6 +74,13 @@ type JobSpec struct {
 	// +optional
 	Completions *int32 `json:"completions,omitempty" protobuf:"varint,2,opt,name=completions"`
 
+	// FailureThreshold specifies the desire to have a job finish after a set number of failures occur.
+	// Setting this value means that the failure of any pod will contribute to the threshold and and signal all pods
+	// to finish once the threshold is met.
+	// More info: http://kubernetes.io/docs/user-guide/jobs
+	// +optional
+	FailureThreshold *int32 `json:"failureThreshold,omitempty" protobuf:"varint,2,opt,name=failureThreshold"`
```

Comment:
Per [this](https://github.com/kubernetes/kubernetes/issues/30243#issuecomment-242871145) and [this](https://github.com/kubernetes/kubernetes/issues/24533#issuecomment-217058534) I'm imaging this allowing you set different policies, to start with:

1. count failures from beginning, iow. I don't care when the failures started, but as soon as they hit X we can fail a job
2. count failures from certain point (probably configurable, eg. X successful), and Y failures after that fails the job
3. be able to specify whether to clear the counter if successful run happens in the mean time

@erictune we've never actually specified how this could be implemented, do you have something on my your mind for this or what I've mentioned could be a good place to start with?